File: Battleships.App/ClientApp/src/components/views/Game.js
import { useState } from 'react';
import { GridCell } from "../grid/GridCell";
import { Grid } from "../grid/Grid";

export function Game() {
    const [game, setGame] = useState(null);
    const [cells, setCells] = useState([]);

    function onGameFinished() {
        alert("Game finished!");
    }

    function onShipSunken(ship) {
        alert(`Ship sunken: ${ship}`);
    }

    async function startGame() {
        const response = await fetch(`api/game`, {
            method: "POST"
        });

        if (!response.ok) {
            throw new Error(`Something went wrong, couldn't start game`);
        }

        const game = await response.json();

        let cells = [];
        for (var y = 0; y < game.gridHeight; y++) {
            for (var x = 0; x < game.gridWidth; x++) {
                cells.push((<GridCell key={x * 10 + y}
                    gameGuid={game.gameGuid}
                    x={x} y={y}
                    onGameFinished={onGameFinished}
                    onShipSunk={onShipSunken}
                />));
            }
        }

        setCells(cells);
        setGame(game);

    }

    return (
        <div>
            <h1>Welcome to my app</h1>
            {game !== null &&
                <Grid sizeX={10} sizeY={10}>
                    {cells.map(cell => cell)}
                </Grid>
            }
            {game === null && <button onClick={startGame}>Start game</button>}
        </div>
    );
}

File: Battleships.App/Persistance/Entities/Game.cs
namespace Battleships.App.Persistance.Entities;

public class Game
{
    public Guid GameGuid { get; }
    public bool Finished { get; set; }
    public int GridWidth { get; }
    public int GridHeight { get; }
    public Dictionary<Position, Cell> Cells { get; }

    public Game(int gridWidth, int gridHeight)
    {
        GameGuid = Guid.NewGuid();
        GridWidth = gridWidth;
        GridHeight = gridHeight;
        Cells = new Dictionary<Position, Cell>();
    }
}
